fms/modules/linear.py

```diff
@@ -0,0 +1,169 @@
+import torch
+import torch.nn as nn
+from typing import Callable, Mapping, Any
+
+
+__type_factory_map: Mapping[str, Callable] = {}
+__type_sharding_map : Mapping[str, Callable] = {}
+
+
+def register_linear_type_to_module_map(linear_type: str, factory: Callable) -> None:
+    if linear_type in __type_factory_map:
+        raise KeyError(
+            f"Module mapping of linear type `{linear_type}` already registered"
+        )
+    __type_factory_map[linear_type] = factory
+
+
+def register_linear_type_to_sharding_map(linear_type: str, factory: Callable) -> None:
+    if linear_type in __type_sharding_map:
+        raise KeyError(
+            f"Sharding map of linear type `{linear_type}` already registered"
+        )
+    __type_sharding_map[linear_type] = factory
+
+
+def get_all_linear_type_to_sharding_maps() -> dict[str, Callable]:
+    return __type_sharding_map
+
+
+def get_linear_type(linear_config: Mapping[str, Any] | None) -> str:
+    if not linear_config:
+        return "torch_linear"
+
+    linear_type = linear_config.get("linear_type", None)
+    if not linear_type:
+        return "torch_linear"
+    if not isinstance(linear_type, str):
+        raise TypeError("linear_type in linear_config must be string")
+    if linear_type.lower() not in __type_factory_map:
+        raise ValueError(f"Unsupported linear_type `{linear_type}` in linear_config")
+
+    return linear_type.lower()
+
+
+def get_linear(
+    in_features: int,
+    out_features: int,
+    bias: bool,
+    linear_config: Mapping[str, Any] | None = None,
+) -> nn.Module:
+    linear_type = get_linear_type(linear_config)
+
+    # TODO: how to merge these calls that get different arguments?
+    if linear_type in __type_factory_map:
+        if linear_type == "torch_linear":
+            return __type_factory_map[linear_type](in_features, out_features, bias)
+        else:
+            return __type_factory_map[linear_type](
+                in_features, out_features, bias, linear_config
+            )
+    else:
+        raise KeyError(f"Unsupported linear type `{linear_type}`")
+
+
+def shard_base_linear(
+    tensor_values: torch.Tensor,
+    self,  # hint should be: type(TPMultiHeadAttention) | type(TPFeedForwardBlock)
+    modules: list[str],
+    params: list[str],
+    name_to_module: dict[str, nn.Module],
+    module_base_shard_dim: dict[str, int],
+    max_partition: dict[str, str],
+    unsharded: list[tuple[str, str]],
+) -> None:
+    all_params = {}
+    used_keys: set[str] = set()
+
+    # Collect quantization parameters to copy on sharded module
+    tensor_device = None
+    for module_name in modules:
+        for param_name in params:
+            if module_name not in all_params:
+                all_params[module_name] = {}
+            # TODO: reusing method '_get_sd_weight' but consider changing its name
+            all_params[module_name][param_name] = self._get_sd_weight(
+                tensor_values, used_keys, [module_name, param_name]
```

Comment:
How does this work? `tensor_values` is not of type Dict

Reply:
Does `_get_sd_weight` load the full tensor or a slice?

Reply:
full tensor, the slice happens at `sharded_copy` as sometimes slices are weird and there's a lot of logic for that